src/storage/gcs/write.c

```diff
@@ -0,0 +1,365 @@
+    ASSERT(bufSize(this->chunkBuffer) > 0);
+    ASSERT(!done || this->uploadId != NULL);
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        // Complete prior async request, if any
+        storageWriteGcsBlock(this, false);
+
+        // Build query
+        HttpQuery *query = httpQueryNewP();
+        httpQueryAdd(query, GCS_QUERY_NAME_STR, strSub(this->interface.name, 1));
+        httpQueryAdd(query, GCS_QUERY_UPLOAD_TYPE_STR, GCS_QUERY_RESUMABLE_STR);
+
+        // Get the upload id
+        if (this->uploadId == NULL)
+        {
+            HttpResponse *response = storageGcsRequestP(this->storage, HTTP_VERB_POST_STR, .upload = true, .query = query);
+
+            MEM_CONTEXT_BEGIN(this->memContext)
+            {
+                this->uploadId = strDup(httpHeaderGet(httpResponseHeader(response), GCS_HEADER_UPLOAD_ID_STR));
+                CHECK(this->uploadId != NULL);
+            }
+            MEM_CONTEXT_END();
+        }
+
+        // Add data to md5 hash
+        ioFilterProcessIn(this->md5hash, this->chunkBuffer);
+
+        // Upload the chunk. If this is the last chunk then add the total bytes in the file to the range rather than the * added to
+        // prior chunks. This indicates that the resumable upload is complete.
+        HttpHeader *header = httpHeaderAdd(
+            httpHeaderNew(NULL), HTTP_HEADER_CONTENT_RANGE_STR,
+            strNewFmt(
+                HTTP_HEADER_CONTENT_RANGE_BYTES " %" PRIu64 "-%" PRIu64 "/%s", this->uploadTotal,
+                this->uploadTotal + bufUsed(this->chunkBuffer) - 1,
+                done ? strZ(strNewFmt("%" PRIu64, this->uploadTotal + bufUsed(this->chunkBuffer))) : "*"));
+
+        httpQueryAdd(query, GCS_QUERY_UPLOAD_ID_STR, this->uploadId);
+
+        // Add fields needed to verify to upload
+        if (done)
+            httpQueryAdd(query, GCS_QUERY_FIELDS_STR, GCS_QUERY_FIELDS_VALUE_STR);
+
+        MEM_CONTEXT_BEGIN(this->memContext)
+        {
+            this->request = storageGcsRequestAsyncP(
+                this->storage, HTTP_VERB_PUT_STR, .upload = true, .noAuth = true, .header = header, .query = query,
+                .content = this->chunkBuffer);
+        }
+        MEM_CONTEXT_END();
+
+        this->uploadTotal += bufUsed(this->chunkBuffer);
+    }
+    MEM_CONTEXT_TEMP_END();
+
+    FUNCTION_LOG_RETURN_VOID();
+}
+
+/***********************************************************************************************************************************
+Write to internal buffer
+***********************************************************************************************************************************/
+static void
+storageWriteGcs(THIS_VOID, const Buffer *buffer)
+{
+    THIS(StorageWriteGcs);
+
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(STORAGE_WRITE_GCS, this);
+        FUNCTION_LOG_PARAM(BUFFER, buffer);
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+    ASSERT(this->chunkBuffer != NULL);
```

Comment:
Shouldn't there be a `ASSERT(buffer != NULL);`? There is in S3 - and I think it somehow got dropped in Azure which we may now be propagating.

Reply:
Fixed for GCS in b6a19f68bd27c8b008. I'll commit to Azure fix to integration directly.